File: frontend/src/lib/adminPermissions.ts
export type AdminAction = "view" | "create" | "edit" | "delete";

export type AdminModule =
  | "overview"
  | "bookings"
  | "promo"
  | "transfer"
  | "content"
  | "fleet"
  | "paths"
  | "vehicles"
  | "users";

export type AdminModulePermissions = Record<AdminAction, boolean>;

export type AdminPermissions = Record<AdminModule, AdminModulePermissions>;

export type AdminSession = {
  username: string;
  role: string;
  is_super_admin: boolean;
  permissions: AdminPermissions;
};

export const ADMIN_SESSION_KEY = "khareef-admin-session";

export const ADMIN_MODULES: AdminModule[] = [
  "overview",
  "bookings",
  "promo",
  "transfer",
  "content",
  "fleet",
  "paths",
  "vehicles",
  "users"
];

export const TAB_MODULE_MAP: Record<string, AdminModule> = {
  overview: "overview",
  bookings: "bookings",
  promo: "promo",
  transfer: "transfer",
  content: "content",
  fleet: "fleet",
  paths: "paths",
  vehicles: "vehicles",
  users: "users"
};

const ACTIONS: AdminAction[] = ["view", "create", "edit", "delete"];

export function emptyPermissions(): AdminPermissions {
  return Object.fromEntries(
    ADMIN_MODULES.map((module) => [module, { view: false, create: false, edit: false, delete: false }])
  ) as AdminPermissions;
}

export function fullPermissions(): AdminPermissions {
  return Object.fromEntries(
    ADMIN_MODULES.map((module) => [module, { view: true, create: true, edit: true, delete: true }])
  ) as AdminPermissions;
}

export function normalizePermissions(raw: Partial<AdminPermissions> | null | undefined): AdminPermissions {
  const base = emptyPermissions();
  if (!raw) return base;
  for (const module of ADMIN_MODULES) {
    const moduleValue = raw[module];
    if (!moduleValue) continue;
    for (const action of ACTIONS) {
      base[module][action] = Boolean(moduleValue[action]);
    }
  }
  return base;
}

export function can(session: AdminSession | null, module: AdminModule, action: AdminAction): boolean {
  if (!session) return false;
  if (session.is_super_admin) return true;
  if (session.role === "normal" && action !== "view") return false;
  return Boolean(session.permissions[module]?.[action]);
}

export function stripNonViewPermissions(permissions: AdminPermissions): AdminPermissions {
  const next = normalizePermissions(permissions);
  for (const module of ADMIN_MODULES) {
    next[module].create = false;
    next[module].edit = false;
    next[module].delete = false;
  }
  return next;
}

export function isSuperAdminSession(session: AdminSession | null): boolean {
  if (!session) return false;
  return session.is_super_admin || session.role === "super_admin";
}

export function canViewTab(session: AdminSession | null, tabId: string): boolean {
  const module = TAB_MODULE_MAP[tabId];
  if (!module) return false;
  if (module === "users") return isSuperAdminSession(session);
  return can(session, module, "view");
}

export function saveAdminSession(session: AdminSession) {
  localStorage.setItem(ADMIN_SESSION_KEY, JSON.stringify(session));
}

export function loadAdminSession(): AdminSession | null {
  try {
    const raw = localStorage.getItem(ADMIN_SESSION_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw) as AdminSession;
    return {
      username: parsed.username || "",
      role: parsed.role || "admin",
      is_super_admin: Boolean(parsed.is_super_admin),
      permissions: normalizePermissions(parsed.permissions)
    };
  } catch {
    return null;
  }
}

export function clearAdminSession() {
  localStorage.removeItem(ADMIN_SESSION_KEY);
}

export function sessionFromAuth(response: {
  username: string;
  role: string;
  is_super_admin?: boolean;
  permissions?: AdminPermissions | null;
}): AdminSession {
  const isSuper = Boolean(response.is_super_admin) || response.role === "super_admin";
  return {
    username: response.username,
    role: response.role,
    is_super_admin: isSuper,
    permissions: isSuper ? fullPermissions() : normalizePermissions(response.permissions || undefined)
  };
}

File: backend/app/admin_permissions.py
"""Admin role and per-module permission helpers."""

from __future__ import annotations

import json
from typing import Any

from fastapi import Depends, HTTPException, status
from sqlalchemy.orm import Session

from . import auth, models
from .database import get_db

ROLE_SUPER_ADMIN = "super_admin"
ROLE_ADMIN = "admin"
ROLE_SCANNER = "scanner"

MODULES = (
    "overview",
    "bookings",
    "promo",
    "transfer",
    "content",
    "fleet",
    "paths",
    "vehicles",
    "users",
)

ACTIONS = ("view", "create", "edit", "delete")

TRANSFER_FIELD_PREFIX = "transfer_"


def full_permissions() -> dict[str, dict[str, bool]]:
    return {module: {action: True for action in ACTIONS} for module in MODULES}


def empty_permissions() -> dict[str, dict[str, bool]]:
    return {module: {action: False for action in ACTIONS} for module in MODULES}


def normalize_permissions(raw: Any) -> dict[str, dict[str, bool]]:
    base = empty_permissions()
    if not isinstance(raw, dict):
        return base
    for module in MODULES:
        module_value = raw.get(module)
        if not isinstance(module_value, dict):
            continue
        for action in ACTIONS:
            base[module][action] = bool(module_value.get(action))
    return base


def permissions_from_admin(admin: models.Admin) -> dict[str, dict[str, bool]]:
    if is_super_admin(admin):
        return full_permissions()
    raw = getattr(admin, "permissions", None)
    if not raw:
        return full_permissions()
    if isinstance(raw, str):
        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError:
            return full_permissions()
        return normalize_permissions(parsed)
    return normalize_permissions(raw)


def permissions_to_json(permissions: dict[str, dict[str, bool]]) -> str:
    return json.dumps(normalize_permissions(permissions), separators=(",", ":"))


def is_super_admin(admin: models.Admin) -> bool:
    role = getattr(admin, "role", None) or ROLE_ADMIN
    return role == ROLE_SUPER_ADMIN


def is_panel_admin(admin: models.Admin) -> bool:
    role = getattr(admin, "role", None) or ROLE_ADMIN
    return role in {ROLE_SUPER_ADMIN, ROLE_ADMIN}


def has_permission(admin: models.Admin, module: str, action: str) -> bool:
    if is_super_admin(admin):
        return True
    permissions = permissions_from_admin(admin)
    module_perms = permissions.get(module, {})
    return bool(module_perms.get(action))


def require_permission(module: str, action: str):
    def _dependency(
        admin: models.Admin = Depends(auth.get_current_admin),
    ) -> models.Admin:
        if not has_permission(admin, module, action):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"You do not have permission to {action} {module}.",
            )
        return admin

    return _dependency


def require_any_permission(modules: tuple[str, ...], action: str):
    def _dependency(
        admin: models.Admin = Depends(auth.get_current_admin),
    ) -> models.Admin:
        if any(has_permission(admin, module, action) for module in modules):
            return admin
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You do not have permission for this action.",
        )

    return _dependency


def require_super_admin():
    def _dependency(
        admin: models.Admin = Depends(auth.get_current_admin),
    ) -> models.Admin:
        if not is_super_admin(admin):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only the main administrator can manage admin users.",
            )
        return admin

    return _dependency


def can_edit_site_content(admin: models.Admin) -> bool:
    return has_permission(admin, "content", "edit") or has_permission(admin, "transfer", "edit")


def filter_site_content_payload(
    admin: models.Admin,
    payload: dict[str, Any],
) -> dict[str, Any]:
    if is_super_admin(admin):
        return payload
    content_edit = has_permission(admin, "content", "edit")
    transfer_edit = has_permission(admin, "transfer", "edit")
    if content_edit and transfer_edit:
        return payload
    allowed: dict[str, Any] = {}
    for key, value in payload.items():
        is_transfer_field = key.startswith(TRANSFER_FIELD_PREFIX)
        if is_transfer_field and transfer_edit:
            allowed[key] = value
        elif not is_transfer_field and content_edit:
            allowed[key] = value
    return allowed


def permissions_to_schema(admin: models.Admin) -> schemas.AdminPermissionsOut:
    from . import schemas

    raw = permissions_from_admin(admin)
    return schemas.AdminPermissionsOut(**raw)


def admin_user_out(admin: models.Admin) -> schemas.AdminUserOut:
    from . import schemas

    return schemas.AdminUserOut(
        id=admin.id,
        username=admin.username,
        role=admin.role,
        is_super_admin=is_super_admin(admin),
        permissions=permissions_to_schema(admin),
        created_at=admin.created_at,
    )


def auth_token_payload(admin: models.Admin) -> dict:
    role = auth.admin_role(admin)
    return {
        "access_token": auth.create_access_token(admin.username, role),
        "token_type": "bearer",
        "role": role,
        "username": admin.username,
        "is_super_admin": is_super_admin(admin),
        "permissions": permissions_to_schema(admin),
    }


def count_super_admins(db: Session, exclude_id: int | None = None) -> int:
    query = db.query(models.Admin).filter(models.Admin.role == ROLE_SUPER_ADMIN)
    if exclude_id is not None:
        query = query.filter(models.Admin.id != exclude_id)
    return query.count()